src/index.js

```diff
@@ -1,15 +1,17 @@
-import VueGapi from './VueGapi'
+import GapiClientProvider from './VueGapi/GapiClientProvider'
+import GoogleAuthService from './VueGapi/GoogleAuthService'
+import { getObjectCopy } from './VueGapi/utils'
 
-function plugin(Vue, clientConfig) {
-  Vue.use(VueGapi, clientConfig)
-}
+export default {
+  install: (Vue, clientConfig) => {
+    const gapiClientProvider = new GapiClientProvider(
+      getObjectCopy(clientConfig)
+    )
 
-// Install by default if using the script tag
-if (typeof window !== 'undefined' && window.Vue) {
-  window.Vue.use(plugin)
+    Vue.prototype.$gapi = new GoogleAuthService(gapiClientProvider)
```

Comment:
@cedpoilly / others, I would love some feedback on this approach which exposes the `GoogleAuthService` instance directly and introduces a `GapiClientProvider` dependency (open to suggestions on naming).

This will most noticeably affect the implementation of `GoogleAuthService` methods which interact with `GapiClientProvider`'s promise-based methods in favor of referencing an `authInstance` property which was set externally. I've updated `login` / `logout` as an example of what this change would look like.

This will likely be a breaking change and could give us a reason to move to Vue v3.x and/or improve APIs while we're at it. Let me know what you think.

Reply:
I prefer a promise based approach on the function rather than onResolve, onReject callbacks. As your approach is a promise based approach, the design changes look good to me ( LGTM! ). Also it's better if we also expose the `authInstancce` to the user as `this.$gapi.getAuthInstance()` as the user will be able to implement their own functionality extending the library. Readme should mention to use the library's login / logout as we do use localStorage for cache mechanism.